map_view/display-device-location-with-nmea-data-sources/src/main/java/com/esri/samples/display_device_location_with_nmea_data_sources/DisplayDeviceLocationWithNmeaDataSourcesController.java

```diff
@@ -0,0 +1,203 @@
+   * on the location display. Data is pushed to the data source using a time line to simulate live updates, as they would
+   * appear if using real-time data from a GPS dongle.
+   */
+  @FXML
+  private void start () {
+
+    // initialize the location data source and prepare to begin receiving location updates when data is pushed. As
+    // updates are received, they will be displayed on the map
+    nmeaLocationDataSource.startAsync();
+
+    // prepare the mock data NMEA sentences
+    File simulatedNmeaDataFile = new File(System.getProperty("data.dir"), "./samples-data/redlands/Redlands.nmea");
+    if (simulatedNmeaDataFile.exists()) {
+
+      try  {
+        // read the nmea file contents using a buffered reader and store the mock data sentences in a list
+        BufferedReader bufferedReader = new BufferedReader(new FileReader(simulatedNmeaDataFile.getPath()));
+        String nmeaSentence;
+        List<String> nmeaSentences = new ArrayList<>();
+
+        while ((nmeaSentence = bufferedReader.readLine()) != null) {
+          // add carriage return for NMEA location data source parser
+          nmeaSentences.add(nmeaSentence + "\n");
+        }
+        // close the stream and release resources
+        bufferedReader.close();
+
+        // add a satellite changed listener to the NMEA location data source and display satellite information on the app
+        setupSatelliteChangedListener();
+
+        // create a new time line to push the mock data NMEA sentences into the data source every 250 ms
+        timeline = new Timeline();
+        timeline.setCycleCount(-1); // loop count
+        timeline.getKeyFrames().add(new KeyFrame(Duration.millis(250), event -> {
+
+          // note: you can also use real-time NMEA sentences obtained via a GPS dongle
+          nmeaLocationDataSource.pushData(nmeaSentences.get(count++).getBytes(StandardCharsets.UTF_8)); // post increment step
+          // reset the count after the last data point is reached
+          if (count == nmeaSentences.size()) count = 0;
+
+        }));
+        // start the time line
+        timeline.play();
+
+        startButton.setDisable(true);
+        stopButton.setDisable(false);
+
+      } catch (Exception e) {
+        new Alert(Alert.AlertType.ERROR, e.getCause().getMessage()).show();
+        e.printStackTrace();
+      }
+
+    } else {
+      new Alert(Alert.AlertType.ERROR, "File not found").show();
+    }
+
+  }
+
+  /**
+   * Obtains NMEA satellite information from the NMEA location data source, and displays satellite information on the app.
+   */
+  private void setupSatelliteChangedListener() {
+
+    HashSet<Integer> uniqueSatelliteIds = new HashSet<>();
+
+    nmeaLocationDataSource.addSatellitesChangedListener(satellitesChangedEvent -> {
+      // get satellite information from the nmea location data source every time the satellites change
```

Comment:
```suggestion

      // get satellite information from the NMEA location data source every time the satellites change
```
Again just a little whitespace here made it easier for me as I read down. Also note nmea -> NMEA.